src/Compilers/CSharp/Portable/Symbols/TypeWithAnnotations.cs

```diff
@@ -919,6 +938,103 @@ internal override void TryForceResolve(bool asValueType)
             }
         }
 
+        /// <summary>
+        /// Extension for a type argument whose nullable annotation after substitution depends on the type constraints on the type parameter.
+        /// To avoid a cycle, we delay determining the nullable annotation until after the type parameter's type constraints are bound.
+        /// </summary>
+        private sealed class LazySubstitutedType : Extensions
+        {
+            private readonly ImmutableArray<CustomModifier> _customModifiers;
+            private readonly TypeParameterSymbol _typeParameter;
+            private int _resolved;
+
+            public LazySubstitutedType(ImmutableArray<CustomModifier> customModifiers, TypeParameterSymbol typeParameter)
+            {
+                Debug.Assert(!customModifiers.IsDefault);
+                _customModifiers = customModifiers;
+                _typeParameter = typeParameter;
+                _resolved = (int)NullableAnnotation.Ignored;
+            }
+
+            internal override SpecialType GetSpecialType(TypeSymbol typeSymbol) => typeSymbol.SpecialType;
+            internal override bool IsRestrictedType(TypeSymbol typeSymbol, bool ignoreSpanLikeTypes) => typeSymbol.IsRestrictedType(ignoreSpanLikeTypes);
+            internal override bool IsStatic(TypeSymbol typeSymbol) => typeSymbol.IsStatic;
+            internal override bool IsVoid(TypeSymbol typeSymbol) => typeSymbol.IsVoidType();
+            internal override bool IsSZArray(TypeSymbol typeSymbol) => typeSymbol.IsSZArray();
+            internal override bool IsRefLikeType(TypeSymbol typeSymbol) => typeSymbol.IsRefLikeType;
+
+            internal NullableAnnotation GetResolvedAnnotationCore()
+            {
+                if (_typeParameter.IsNotNullable == true)
+                {
+                    return NullableAnnotation.NotAnnotated;
+                }
+                else
+                {
+                    return NullableAnnotation.Oblivious;
```

Comment:
nit: May deserve a comment (we only use this lazy extension for oblivious type parameter scenario). Alternatively, we could add "Oblivious" in the name of the type, but that might be less clear